skfs/src/EagerDir.sk

```diff
@@ -1576,105 +1577,130 @@ class EagerDir{
               !ranges.any(range -> range.start <= key && key <= range.end)
             ) ->
               void
-            | _ -> !dirty = dirty.set(key)
+            | _ -> !dirtyInRegion = dirtyInRegion.set(key)
             }
-          }
-        }
-      };
+          };
+          dirtyInRegion;
+        })
+      }
     };
     if (contextDirty.containsKey(parent.dirName)) {
-      for (key in contextDirty[parent.dirName]) {
+      keys = contextDirty[parent.dirName].values();
+      !dirty = withRegionIterate(keys, dirty, (key, dirtyInRegion) ~> {
         for (p in parentMaps) {
           (_, rangeOpt, _) = p;
           rangeOpt match {
           | Some(
             ranges,
           ) if (!ranges.any(range -> range.start <= key && key <= range.end)) ->
             void
-          | _ -> !dirty = dirty.set(key)
+          | _ -> !dirtyInRegion = dirtyInRegion.set(key)
           }
-        }
-      };
-    };
-
-    for (key in dirty) {
-      context.enter(ArrowKey(parentName, childName, key));
-      path = Path::create(parentName, key);
-      source = Source::create(path);
-      oldInfo = static::getOld(child, source);
-      oldKeys = SortedSet::createFromItems(oldInfo.getKeys());
-      newDirsCopy = context.newDirs;
-      readsCopy = context.reads;
-      context.!newDirs = SortedSet[];
-      context.!reads = SortedSet[];
-
-      keys = mutable Vector[];
-      mvalues = SortedMap[];
-
-      for (p in parentMaps) {
-        writer = mutable Writer{};
-        (mapFun, _, _) = p;
-        mapFun(context, writer, key, parent.getIterRaw(key));
-        mapped = writer.getWrites();
-        for (kv in mapped) {
-          (k, rvalues) = kv;
-
-          keys.push(k);
-          if (oldKeys.contains(k)) {
-            !oldKeys = oldKeys.remove(k);
-          };
+        };
+        dirtyInRegion;
+      })
+    };
+
+    it = dirty.iterator();
+    dKeys = Array<Key>::createFromIterator(it);
+    index = 0;
+    size = dKeys.size();
+    while (index < size) {
+      indexCopy = index;
+      childCopy = childRef;
+      (!index, !childRef) = withRegion(context, (obstack, context) ~> {
+        idx = indexCopy;
+        child = childCopy;
```

Comment:
I haven't thought through in detail, but would it be possible to simplify this dance between e.g. `index`, `indexCopy`, and `idx` using your new `withRegionIterate`?